File: core/schemas/edl.py
"""
EDL / timeline domain model — pydantic mirror of ``shared/types/edl.ts``.

Conventions (kept identical to the TS side):
  - all times are SECONDS (float) on the project clock;
  - intervals are half-open ``[startSec, endSec)``;
  - ids are opaque strings.

Field names use the TS camelCase spelling so JSON round-trips through the Node
gateway to the frontend editor without any remapping.
"""

from __future__ import annotations

from typing import Literal, Optional, Union

from pydantic import BaseModel, Field


class TimeRange(BaseModel):
    startSec: float
    endSec: float

    @property
    def duration(self) -> float:
        return max(0.0, self.endSec - self.startSec)


class Word(BaseModel):
    text: str
    startSec: float
    endSec: float
    confidence: Optional[float] = None


class Transcript(BaseModel):
    text: str
    words: list[Word] = Field(default_factory=list)
    language: str = "en"
    durationSec: float = 0.0


class BeatQueries(BaseModel):
    said: str
    shown: str
    keywords: list[str] = Field(default_factory=list)


class Beat(BaseModel):
    id: str
    text: str
    range: TimeRange
    queries: Optional[BeatQueries] = None
    # Planned visual treatment (Ch11 scene planning), decided BEFORE retrieval:
    # stock_video | stock_image | ai_image | motion_graphics
    visual: Optional[str] = None


class ClipAsset(BaseModel):
    id: str
    path: str
    durationSec: float
    width: int
    height: int
    fps: Optional[float] = None
    tags: list[str] = Field(default_factory=list)
    thumbPath: Optional[str] = None
    source: Literal["user", "stock"] = "stock"
    license: Optional[str] = None


class GenerationSlot(BaseModel):
    id: str
    beatId: str
    prompt: str
    durationSec: float
    status: Literal["pending", "generating", "done", "failed"] = "pending"
    assetId: Optional[str] = None


class AssetClipSource(BaseModel):
    kind: Literal["asset"] = "asset"
    assetId: str
    inSec: float
    outSec: float


class GenerateClipSource(BaseModel):
    kind: Literal["generate"] = "generate"
    slot: GenerationSlot


ClipSource = Union[AssetClipSource, GenerateClipSource]


class TimelineClip(BaseModel):
    id: str
    beatId: Optional[str] = None
    source: ClipSource
    range: TimeRange
    label: Optional[str] = None
    review: Optional[bool] = None
    matchScore: Optional[float] = None
    # Look preset applied to this clip: `lookId` labels it in the editor,
    # `look` is the ffmpeg filter chain the exporter appends.
    lookId: Optional[str] = None
    look: Optional[str] = None
    # Editing Lab recipe this clip was composed from (preset id + control
    # values), so the editor can re-render the same shot with new settings.
    shotSpec: Optional[dict] = None
    # Playback gain for audio-track clips (1.0 = unity).
    gain: Optional[float] = None


TrackKind = Literal["video", "overlay", "audio", "captions"]


class Track(BaseModel):
    id: str
    kind: TrackKind
    name: str
    clips: list[TimelineClip] = Field(default_factory=list)
    muted: Optional[bool] = None
    locked: Optional[bool] = None


class CaptionCue(BaseModel):
    id: str
    text: str
    range: TimeRange


class Timeline(BaseModel):
    id: str
    fps: float = 30
    width: int = 1920
    height: int = 1080
    durationSec: float = 0.0
    audioPath: Optional[str] = None
    tracks: list[Track] = Field(default_factory=list)
    captions: list[CaptionCue] = Field(default_factory=list)


class Project(BaseModel):
    id: str
    title: str
    createdAt: str
    updatedAt: str
    timeline: Timeline
    transcript: Optional[Transcript] = None
    beats: Optional[list[Beat]] = None

File: shared/types/edl.ts
/**
 * EDL (edit decision list) / timeline domain types — the editor's data model.
 *
 * Conventions:
 *  - All times are in SECONDS (floating point) on the project clock.
 *  - All ids are opaque strings (nanoid-style).
 */

/** A half-open interval [startSec, endSec) on the project clock. */
export interface TimeRange {
  startSec: number;
  endSec: number;
}

/** One transcribed word with its timing. */
export interface Word {
  text: string;
  startSec: number;
  endSec: number;
  /** Confidence 0..1 when the ASR engine provides one. */
  confidence?: number;
}

/** Full transcript of the narration audio. */
export interface Transcript {
  text: string;
  words: Word[];
  language: string;
  /** Duration of the source audio in seconds. */
  durationSec: number;
}

/** A "beat" — the smallest visual unit of the story (one clip per beat). */
export interface Beat {
  id: string;
  /** The narration text spoken during this beat. */
  text: string;
  /** Position of the beat on the audio clock. */
  range: TimeRange;
  /** Search queries (what's SAID vs what's SHOWN). */
  queries?: BeatQueries;
}

/** Retrieval queries for one beat. */
export interface BeatQueries {
  /** Semantic query for the narration content (what is said). */
  said: string;
  /** Visual query describing the desired imagery (what is shown). */
  shown: string;
  /** Optional extra keyword terms. */
  keywords?: string[];
}

/** A clip in the local library (user footage or downloaded free stock). */
export interface ClipAsset {
  id: string;
  /** Storage-relative path to the media file. */
  path: string;
  durationSec: number;
  width: number;
  height: number;
  fps?: number;
  /** Freeform descriptive tags (used for filtering, not ranking). */
  tags: string[];
  /** Path to a poster/thumbnail frame, if extracted. */
  thumbPath?: string;
  /** Where the clip came from. */
  source: 'user' | 'stock';
  /** License note for stock footage (e.g. "Pexels", "CC0"). */
  license?: string;
}

/**
 * A slot on the timeline the pipeline could not fill from the library —
 * rendered as a placeholder so a generator can fill it later.
 */
export interface GenerationSlot {
  id: string;
  beatId: string;
  /** Text-to-video prompt derived from the beat's `shown` query. */
  prompt: string;
  /** Desired clip length in seconds. */
  durationSec: number;
  status: 'pending' | 'generating' | 'done' | 'failed';
  /** Populated once a generator produced an asset. */
  assetId?: string;
}

/** Source of a timeline clip: an existing asset, or a to-be-generated slot. */
export type ClipSource =
  | {
      kind: 'asset';
      assetId: string;
      /** In/out points INSIDE the source asset, in seconds. */
      inSec: number;
      outSec: number;
    }
  | {
      kind: 'generate';
      slot: GenerationSlot;
    };

/** One clip placed on a track. */
export interface TimelineClip {
  id: string;
  /** The beat this clip covers (empty for manually added clips). */
  beatId?: string;
  source: ClipSource;
  /** Position on the project clock. */
  range: TimeRange;
  /** Optional user label shown in the editor. */
  label?: string;
  /** Flagged (yellow outline) when a match scored below threshold. */
  review?: boolean;
  /** Match score (0..1 cosine) recorded by the matching pipeline. */
  matchScore?: number;
  /** Playback gain for audio clips (1 = unity). */
  gain?: number;
}

export type TrackKind = 'video' | 'overlay' | 'audio' | 'captions';

/** An ordered lane of non-overlapping clips. */
export interface Track {
  id: string;
  kind: TrackKind;
  name: string;
  clips: TimelineClip[];
  muted?: boolean;
  locked?: boolean;
}

/** A caption cue (burned in or exported as sidecar). */
export interface CaptionCue {
  id: string;
  text: string;
  range: TimeRange;
  /** Per-word timings when the transcript had them (drives word-by-word styles). */
  words?: Word[];
}

/** The full timeline / EDL for one project. */
export interface Timeline {
  id: string;
  fps: number;
  width: number;
  height: number;
  durationSec: number;
  /** Path to the narration audio the timeline is aligned to. */
  audioPath?: string;
  tracks: Track[];
  captions: CaptionCue[];
  /** Burn-in caption style; ids match core/agents/exporter/caption_styles.py. */
  captionStyle?: CaptionStyleId;
  /** Per-project overrides on top of that style. */
  captionOptions?: CaptionOptions;
}

/** Caption burn-in styles offered in the editor. */
export type CaptionStyleId =
  | 'classic' | 'outline' | 'pop' | 'banner' | 'minimal' | 'top'
  | 'live' | 'karaoke';

/** User tweaks layered over the chosen style preset. */
export interface CaptionOptions {
  /** Text height as a percentage of the frame (e.g. 5 = 5% of height). */
  sizePct?: number;
  /** Hex fill, e.g. "#FFE14D". */
  color?: string;
  place?: 'bottom' | 'middle' | 'top';
  upper?: boolean;
}

export const CAPTION_STYLES: { id: CaptionStyleId; label: string; hint: string }[] = [
  { id: 'classic', label: 'Classic', hint: 'White on a soft black box' },
  { id: 'outline', label: 'Bold outline', hint: 'Large white with a thick outline' },
  { id: 'pop', label: 'Yellow pop', hint: 'Impact caps in yellow' },
  { id: 'banner', label: 'Banner', hint: 'Wide dark bar across the bottom' },
  { id: 'minimal', label: 'Minimal', hint: 'Small white with a soft shadow' },
  { id: 'top', label: 'Top bar', hint: 'Boxed caption at the top' },
  { id: 'live', label: 'Live typing', hint: 'Words appear as they are spoken' },
  { id: 'karaoke', label: 'One word', hint: 'One big word at a time, centre screen' },
];

/** A saved editor project (timeline + provenance). */
export interface Project {
  id: string;
  title: string;
  createdAt: string;
  updatedAt: string;
  timeline: Timeline;
  transcript?: Transcript;
  beats?: Beat[];
  /** The pipeline run this project came from — enables "fill missing footage". */
  runId?: string;
}
